src/main/scala/com/amazon/deequ/analyzers/Distance.scala

```diff
@@ -40,32 +52,198 @@ object Distance {
       selectMetrics(linfSimple, n, m, correctForLowNumberOfSamples)
     }
 
-    /** Calculate distance of categorical profiles based on L-Infinity Distance */
-    def categoricalDistance(
-      sample1: scala.collection.mutable.Map[String, Long],
-      sample2: scala.collection.mutable.Map[String, Long],
-      correctForLowNumberOfSamples: Boolean = false)
+  /** Calculate distance of categorical profiles based on different distance methods
+   *
+   * Thresholds for chi-square method:
+   *    - for 2x2 tables: all expected counts should be 10 or greater (Cochran, William G. "The χ2 test of goodness of fit." The Annals of mathematical statistics (1952): 315-345.)
+   *    - for tables larger than 2 x 2: "No more than 20% of the expected counts are less than 5 and all individual expected counts are 1 or greater" (Yates, Moore & McCabe, 1999, The Practice of Statistics, p. 734)
+   *
+   * @param sample1                      the mapping between categories(keys) and counts(values) of the observed sample
+   * @param sample2                      the mapping between categories(keys) and counts(values) of the expected baseline
+   * @param correctForLowNumberOfSamples if true returns chi-square statistics otherwise p-value
+   * @param method                       Method to use: LInfinity or Chisquare
+   * @param absThresholdYates            Yates absolute threshold for tables larger than 2x2
+   * @param percThresholdYates           Yates percentage of categories that can be below threshold for tables larger than 2x2
+   * @param absThresholdCochran          Cochran absolute threshold for 2x2 tables
+   * @return distance                    can be an absolute distance or a p-value based on the correctForLowNumberOfSamples argument
+   */
+  def categoricalDistance(
+    sample1: scala.collection.mutable.Map[String, Long],
+    sample2: scala.collection.mutable.Map[String, Long],
+    correctForLowNumberOfSamples: Boolean = false,
+    method: CategoricalDistanceMethod = LInfinity,
+    absThresholdYates: Integer = 5,
+    percThresholdYates: Double = 0.2,
+    absThresholdCochran: Integer = 10)
```

Comment:
It appears that these there additional parameters are not used when the method provided is `LInfinity`. Is it possible to use a case class for each type of method, instead of an enum? That way, we can encapsulate the additional configuration for each method into its case class.

Reply:
refactored to:  ```
 trait CategoricalDistanceMethod
    case class LInfinity() extends CategoricalDistanceMethod
    case class Chisquare(
                          absThresholdYates: Integer = defaultAbsThresholdYates,
                          percThresholdYates: Double = defaultPercThresholdYates,
                          absThresholdCochran: Integer = defaultAbsThresholdCochran)
      extends CategoricalDistanceMethod
```